### ISSUE TITLE
request: implement `@command` decorator

### ISSUE THREAD

**Issue by i0bs** [retuxteam/retux#47]
### What's the feature?

Commands are the cornerstone of every bot on Discord; paramount to their functionality and purpose. For the longest time, bot developers have been trying to create the most accessible commands for their users, but never were there decent solutions for how to do this with the latest slash commands. Similar to the the design of interactions.py's command system, I would like to propose my solution for end-user facing code.

### How does it work?

Now that we've explained the "what" of the decorator, let's explain the "how."

## Motive

In retux, versatility and accessibility are two of the philosophies that we associate with the design and structure of the codebase. Slash commands should be able to achieve numerous methodologies for doing something, simple to use, understand, and easy to work with.

This is the bare functionality, without the usage of classes or other command functionalities. The command decorator will act as a Gateway event handler *purely* for parsing `INTERACTION_CREATE` events. This means that any asynchronous task we register in our bot as a callback will only be triggered when the right conditions are met.

## Design

By base functionality, no data supplied to the decorator will force the library to assume you're creating a chat-input command with no description, options or pre-set permissions.

```py
# Commands are registered via. the use of a coroutine or asynchronous task.
@bot.command
async def test(interaction: retux.Interaction) -> retux.InteractionResponse:
    # Let's note here two huge things about this: the argument signature and return type.
    # Because Discord's API only provides us back an interaction object, it's smarter for us
    # to use this to be more strict with the API instead of creating a custom-made context.
    
    # The second thing is the return type being an InteractionResponse object, something
    # that only the client gives back and allows bot developers to store the response if they
    # want to use it in another command handler.
    
    # Everything for responding to/with interactions is asynchronously called. Content is the only
    # positional argument in the .respond() signature, with everything being keyword-only.
    await interaction.respond("Hello world!")

    # When we make a return, the last response of the command handler will be passed as the
    # interaction response. However, the InteractionResponse object acts as a manager class for
    # data relevant to your response, including past responses made.
    return await interaction.respond(embeds=retux.Embed(title="test"))
```

The lack of a `Context` contextualisation class for abstracting invocation further is admittedly a huge learning curve for developers working from text-based commands to slash commands. I believe it is necessary that we make this change to have a better standard with keeping certain objects purely related to the Gateway's output.

### Guild-specific commands

But let's say we don't want to create commands on Discord's global scope. That's fine, because we have functionality for better parameter definition.

```py
GUILD_ID = retux.Snowflake(1234567890) # str/int input.
GUILD = retux.Guild.get(1234567890) # str/int/Snowflake input.

# We can supply guild IDs (eliminating the global scope) as either a string, integer, Snowflake or Guild object.
@bot.command(guild_ids=["1234567890", 1234567890, GUILD_ID, GUILD])
```

### Command options

Here comes the shining part of the command decorator: the versatility and safety in defining commands.

```py
# This method here allows you to define it with an option decorator. The specifics for how this will work have
# not yet been determined, but its argument signature has one of many entry points.
@bot.command
@bot.option("the option's description", required=False)
async def test(interaction, option_name: str):
    # With this method, we're making the assumption that the option decorator
    # has a linear correlation with the argument signature of our coroutine.
    # This means the sequential order of arguments are relied on for whether they're required
    # or not and their description. Note that the "required" kwarg of the decorator may be 
    # alternatively instituted by making the argument "option_name" have an optional value.
    return await interaction.respond(option_name)
```

Another way of doing it is confining the information of the option to be inside the argument signature of the command handler's coroutine. This will be linearly filled as well in retux to the first method, but the Gateway dispatcher will properly handle
which ones are inputted.

```py
@bot.command
async def test(interaction, option_name: retux.Option[str, "the option's description"] = False):
    return await interaction.respond(option_name) # sends "False" or the given input.
```

Now, it's worth also noting that the command decorator will have the ability to define options within itself as well, similar to how interactions.py does it.

```py
@bot.command(options=[
    ...,
    {
        "type": str,
        "name": "option_name",
        "description": "desc.",
        "required": False
    }, 
    retux.Option(...)
])
async def test(interaction, option_name):
    ...
```

### Custom-defined names

You've also been seeing so far that the name of the command is determined by the name of the coroutine/async task. I can happily say that we do plan to allow `on`-like decorator functionality for the name, where it may be assignable as `@bot.command(name="test")` for custom-named coroutines.

### Code of Conduct

- [X] I agree.

**Comment by Dworv** [comment]
`@bot.command` could be a dual-way decorator like `@bot.on`, allowing command structures like this:
(let me know if you think this is a good idea)
```py
@bot.command
class Member: # these would automatically be set to lowercase, so that users can follow PEP8 and avoid warnings
    """A base description that is not used"""
    @bot.subcommand
    async def admin(self, interaction):
        ...

    @bot.subcommand_group # something like this could also be implemented, also being a dual-way operator and allowing even more structuring for subcommand groups
    class Remove:
        """Remove a member from the server"""
        @bot.subcommand
        async def kick(self, interaction):
            """Kick a member"""
            ....
```

**Comment by i0bs** [comment]
I personally dislike the way we're nesting classes for this, although I acknowledge that this is a Pythonic standard in many deps. This could work if our command decorator was global (e.g. not inside `retux.Bot`) but currently it is not. We'd need to have some global decorator to handle converting these to managers so that our Gateway dispatch can properly run the callback in the Bot client. This would likely be something as `@retux.class.command` or of the like.

Really cool idea though for simplifying the control flow of subcommands into an object orientation. I'd play around with this proposal some more to think of ways to make it better fitting.

**Comment by Dworv** [comment]
Why would it only work if the `@command` decorator was global? Don't get me wrong, I get that it would be nice to have, but I don't see what's stopping us using the `@bot.command` decorator for this.

**Comment by i0bs** [comment]
`@bot.command` will register only for that coroutine it is attached to. If we want to standardise the control flow and hierarchal execution of the class, You can't attach `@bot.subcommand` without making that coroutine independent. Having something more global (and properly handling the argument signature to detect `self`) allows us to generate command managers so that we don't have the hassle of asynchronously calling these callable methods when `self` is missing.

In `retux.Bot._trigger()`, we pass `*args` for numerous positional arguments, which may include `self`. Having a global decorator allows more versatility and influence on controlling the generated arguments required when `GatewayClient._dispatch()` calls the said trigger method. 

I feel that `retux.Bot` should exclusively be for managing and controlling how callbacks are registered; and triggered. Decorators are merely the bot developer's entry point for creating and running a callback. Otherwise, if we kept them in `retux.Bot`, we will (eventually and surely) reach the messiness of interactions.py. in decorator source code.

**Comment by Dworv** [comment]
Alright, so when users use the global decorator, will we just scrape them from the file like NAFF does (afaik)?

**Comment by i0bs** [comment]
No, scraping is bad. All we need the global decorators to do is convert everything into a workable manager class. NAFF's version of a command manager is `BaseCommand` which contains the inherent calling logic for execution. In practice, we'd still have decorators in `retux.Bot`, but they'd do nothing more than point to the globally defined ones. We just need something to work within scopes so that things are still being converted, but not restricted to the client itself.
